File: lib/api/visualizations.ts
import { apiFetch } from "./client"
import type { Visualization, VisualizationConfig } from "../types"

function mapVisualization(row: Record<string, unknown>): Visualization {
  return {
    id: row.id as string,
    programId: row.program_id as string,
    name: row.name as string,
    description: row.description as string | undefined,
    config: row.config as VisualizationConfig,
    createdAt: row.created_at as string,
  }
}

export async function listVisualizations(programId: string): Promise<Visualization[]> {
  const data = await apiFetch<Record<string, unknown>[]>(
    `/api/programs/${programId}/visualizations`
  )
  return data.map(mapVisualization)
}

export async function getVisualization(
  programId: string,
  visualizationId: string
): Promise<Visualization> {
  const data = await apiFetch<Record<string, unknown>>(
    `/api/programs/${programId}/visualizations/${visualizationId}`
  )
  return mapVisualization(data)
}

export async function createVisualization(
  programId: string,
  payload: { name: string; description?: string; config: VisualizationConfig }
): Promise<Visualization> {
  const data = await apiFetch<Record<string, unknown>>(
    `/api/programs/${programId}/visualizations`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        name: payload.name,
        description: payload.description ?? "",
        config: payload.config,
      }),
    }
  )
  return mapVisualization(data)
}

File: backend/routes/visualizations.py
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.db_models import ProgramDB, VisualizationDB
from backend.models.visualization import Visualization, VisualizationCreate

router = APIRouter(prefix="/api/programs/{program_id}/visualizations", tags=["visualizations"])


@router.get("", response_model=list[Visualization])
async def list_visualizations(program_id: str, db: AsyncSession = Depends(get_db)):
    program = await db.execute(select(ProgramDB).where(ProgramDB.id == program_id))
    if program.scalars().first() is None:
        raise HTTPException(status_code=404, detail="Program not found")

    result = await db.execute(
        select(VisualizationDB)
        .where(VisualizationDB.program_id == program_id)
        .order_by(VisualizationDB.created_at.desc())
    )
    return result.scalars().all()


@router.post("", response_model=Visualization, status_code=201)
async def create_visualization(
    program_id: str, payload: VisualizationCreate, db: AsyncSession = Depends(get_db)
):
    program = await db.execute(select(ProgramDB).where(ProgramDB.id == program_id))
    if program.scalars().first() is None:
        raise HTTPException(status_code=404, detail="Program not found")

    row = VisualizationDB(
        id=str(uuid.uuid4()),
        program_id=program_id,
        name=payload.name,
        description=payload.description,
        config=payload.config,
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    db.add(row)
    await db.commit()
    return row


@router.get("/{visualization_id}", response_model=Visualization)
async def get_visualization(
    program_id: str, visualization_id: str, db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(VisualizationDB)
        .where(VisualizationDB.program_id == program_id)
        .where(VisualizationDB.id == visualization_id)
    )
    row = result.scalars().first()
    if not row:
        raise HTTPException(status_code=404, detail="Visualization not found")
    return row
